vm/src/stdlib/itertools.rs

```diff
@@ -181,6 +184,141 @@ impl PyItertoolsTakewhile {
     }
 }
 
+#[pyclass(name = "islice")]
+#[derive(Debug)]
+struct PyItertoolsIslice {
+    iterable: PyObjectRef,
+    cur: RefCell<usize>,
+    next: RefCell<usize>,
+    stop: Option<usize>,
+    step: usize,
+}
+
+impl PyValue for PyItertoolsIslice {
+    fn class(vm: &VirtualMachine) -> PyClassRef {
+        vm.class("itertools", "islice")
+    }
+}
+
+fn pyobject_to_usize(obj: PyObjectRef, vm: &VirtualMachine) -> Option<usize> {
+    let is_int = objtype::isinstance(&obj, &vm.ctx.int_type());
+    if is_int {
+        objint::get_value(&obj).to_usize()
+    } else {
+        None
+    }
+}
+
+#[pyimpl]
+impl PyItertoolsIslice {
+    #[pymethod(name = "__new__")]
+    fn new(_cls: PyClassRef, args: PyFuncArgs, vm: &VirtualMachine) -> PyResult {
+        let (iter, start, stop, step) = match args.args.len() {
+            0 | 1 => {
+                return Err(vm.new_type_error(format!(
+                    "islice expected at least 2 arguments, got {}",
+                    args.args.len()
+                )))
+            }
+
+            2 => {
+                let (iter, stop): (PyObjectRef, PyObjectRef) = args.bind(vm)?;
+                let stop = if !stop.is(&vm.get_none()) {
+                    let stop = pyobject_to_usize(stop, &vm);
+                    if !stop.is_some() {
+                        return Err(vm.new_value_error("Stop argument for islice() must be None or an integer: 0 <= x <= sys.maxsize.".to_string()));
+                    }
+                    stop
+                } else {
+                    None
+                };
+
+                (iter, 0usize, stop, 1usize)
+            }
+            _ => {
+                let (iter, start, stop, step): (
+                    PyObjectRef,
+                    PyObjectRef,
+                    PyObjectRef,
+                    PyObjectRef,
+                ) = args.bind(vm)?;
+                let stop = if !stop.is(&vm.get_none()) {
```

Comment:
The entire `if` statement can be refactored to a `pyobject_to_opt_usize` function, that returns a `PyResult<Option<usize>>`, then you can reuse it a few times.

Reply:
Not sure it will allow that much of code reuse; The tristate usize/`None`/`Err` has to be checked across all sites anyway, and has to be handled differently (Different default values for `None`, different errors for `Err`...)